### PR TITLE
Added Dynamic Shortcuts

### LINKED ISSUE #923: Add static shorts for app
**Feature Request:**
I want to introduce a feature of adding static shorts to the app for activities in navigation drawer for quick access to the users like:
- Recent Transactions
- Charges
- Third Party Transfers

**Feature description**
App shortcuts were introduced in API 25 for the users do their common tasks quickly. using these shortcuts will be more useful. you can visit this [link](https://developer.android.com/guide/topics/ui/shortcuts/#static) for more information.

**Screenshots:**

static

### PR DESCRIPTION
**Pull request by nirmalhk7**
Added dynamic shortcuts, visisble only if viewed from device having Android newer than Nougat 7.1. Other devices will be unharmed.
I've chosen 4 shortcuts here, please let me know if I should choose 4 different ones.

Signed-off-by: Nirmal Khedkar <nirmalhk7@gmail.com>

Fixes #923 

Please Add Screenshots If there are any UI changes.
untitled

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Apply the `AndroidStyle.xml` style template to your code in Android Studio.
- [x] Run the unit tests with `./gradlew check` to make sure you didn't break anything
- [x] If you have multiple commits please combine them into one commit by squashing them.

### CODE REVIEW

**Comment by devansh-299** [comment]
@nirmalhk7 these shortcuts introduce a very serious problem of bypassing the verification steps while opening the app through them. I also faced similar problem while trying to add shortcuts

**Comment by nirmalhk7** [comment]
@devansh-299 Thanks for pointing out! I totally forgot about them. Bypassing security would indeed be quite a serious bug.

Since I'll be doing these dynamically, my idea to solve it is to let the shortcuts all point at `SplashActivity` (which points to `PasscodeActivity`). I'll save the option menu needed using `setAction()` function (the string will be the name of the menu).

So once you click a shortcut, you'd be directed to the `SplashActivity`, from there the `PasscodeActivity` and from there the `HomeActivity`. Before loading the `HomeOldFragment`, we check if `getAction()` has our menus (I've noticed that if it doesnt, its default value is `_android.intent.action.MAIN_`). If it does, we load our option as if it's been loaded from the Navigation Drawer, else we load the `HomeOldFragment`.

I've implemented all that I've set above, and in my testing, it works well. I'll fix the build errors, clean up the code and will push by tonight.

**Comment by nirmalhk7** [comment]
Please review this PR.

Here is the final output (the lag is because I'm using Emulator):

untitled

File: app/src/main/java/org/mifos/mobile/ui/activities/HomeActivity.java [hunk 1]
```diff
@@ -91,12 +105,20 @@ public void onCreate(Bundle savedInstanceState) {
         setContentView(R.layout.activity_home);
         ButterKnife.bind(this);
         clientId = preferencesHelper.getClientId();
-
+        initialiseShortcuts();
         setupNavigationBar();
         setToolbarElevation();
         setToolbarTitle(getString(R.string.home));
+
+        Log.d("NIRMALHK7","HomeActiv "+getIntent().getAction());
```
Comment (devansh-299) [inline app/src/main/java/org/mifos/mobile/ui/activities/HomeActivity.java]: you can remove this once you have tested

File: app/src/main/java/org/mifos/mobile/ui/activities/HomeActivity.java [hunk 2]
```diff
@@ -143,12 +166,70 @@ protected void onResume() {
...
+                    new ShortcutInfo.Builder(this, "second_shortcut")
+                            .setShortLabel(getString(R.string.recent_transactions))
+                            .setLongLabel(getString(R.string.recent_transactions))
+                            .setIcon(Icon.createWithResource(this, R.drawable.ic_label_blue_24dp))
+                            .setIntent(homeIntent.setAction(
+                                    getString(R.string.recent_transactions)))
+                            .build()
+            );
+            shortcutList.add(
+                    new ShortcutInfo.Builder(this, "forth_shortcut")
```
Comment (devansh-299) [inline app/src/main/java/org/mifos/mobile/ui/activities/HomeActivity.java]: create string resources for "forth_shortcut"  and others

File: app/src/main/java/org/mifos/mobile/ui/activities/HomeActivity.java [hunk 3]
```diff
@@ -143,12 +166,70 @@ protected void onResume() {
...
+                            .setShortLabel("3rd Party Transfer.")
```
Comment (devansh-299) [inline app/src/main/java/org/mifos/mobile/ui/activities/HomeActivity.java]: Same here, use string resource files  wherever necessary

File: app/src/main/java/org/mifos/mobile/ui/activities/PassCodeActivity.java [hunk 4]
```diff
@@ -16,10 +17,14 @@
 import org.mifos.mobile.utils.Toaster;
 
 public class PassCodeActivity extends MifosPassCodeActivity {
-
+    private String actions;
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
+        if(getIntent().getAction()!=null && !getIntent().getAction().trim().equals("android.intent.action.MAIN"))
```
Comment (devansh-299) [inline app/src/main/java/org/mifos/mobile/ui/activities/PassCodeActivity.java]: You can create a constant string inside  `util/Constants.java` for "android.intent.action.MAIN"

File: app/src/main/java/org/mifos/mobile/ui/activities/SplashActivity.java [hunk 5]
```diff
@@ -26,9 +40,15 @@ protected void onCreate(Bundle savedInstanceState) {
         if (!passcodePreferencesHelper.getPassCode().isEmpty()) {
             intent = new Intent(this, PassCodeActivity.class);
             intent.putExtra(Constants.INTIAL_LOGIN, true);
+            Log.d("NIRMALHK7","SplashAct "+getIntent().getAction());
```
Comment (devansh-299) [inline app/src/main/java/org/mifos/mobile/ui/activities/SplashActivity.java]: please remove this

**Comment by devansh-299** [comment]
@nirmalhk7 nice work!
You have 3 checkstyle violations because of which your build is failing the tests
You can checkout the violations here `../mifos-mobile/app/build/reports/checkstyle/checkstyle.html`

**Comment by nirmalhk7** [comment]
Just fixed all your above suggestions. Please review the latest push. I get a build error for reasons I cant understand ` error: cannot find symbol import org.mifos.mobile.injection.component.DaggerActivityComponent;
`. I've asked on some of the gitter channels and people suggested me to do "Clean Project", which hasn't really helped a lot. 

Regardless, I've pushed the code. If theres anything I could do about build, let me know. Thanks!

File: app/src/main/java/org/mifos/mobile/ui/activities/SplashActivity.java [hunk 6]
```diff
@@ -26,9 +40,14 @@ protected void onCreate(Bundle savedInstanceState) {
...
             intent = new Intent(this, PassCodeActivity.class);
             intent.putExtra(Constants.INTIAL_LOGIN, true);
+            if (getIntent().getAction() != null &&
+                    !getIntent().getAction().trim().equals(getString(R.string.defaultAction))) {
+                intent.setAction(getIntent().getAction());
+            }
         } else {
             intent = new Intent(this, LoginActivity.class);
         }
+
```
Comment (iamsh4shank) [inline app/src/main/java/org/mifos/mobile/ui/activities/SplashActivity.java]: Is this space required?

File: app/src/main/res/values/strings.xml [hunk 7]
```diff
@@ -343,6 +343,8 @@
     <string name="delete">Delete</string>
     <string name="beneficiary_deleted_successfully">Beneficiary deleted successfully</string>
     <string name="third_party_transfer">Third Party Transfer</string>
+    <string name="third_party_transfer_short">3rd Party Transfer</string>
+
```
Comment (iamsh4shank) [inline app/src/main/res/values/strings.xml]: Here also

File: app/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -610,5 +612,10 @@
         <item>ru</item>
         <item>sw</item>
     </string-array>
+    <string name="defaultAction">android.intent.action.MAIN</string>
```
Comment (iamsh4shank) [inline app/src/main/res/values/strings.xml]: After string array leave a space of one line.

File: app/src/main/java/org/mifos/mobile/ui/activities/PassCodeActivity.java [hunk 9]
```diff
@@ -16,10 +17,14 @@
 import org.mifos.mobile.utils.Toaster;
 
 public class PassCodeActivity extends MifosPassCodeActivity {
-
+    private String actions;
```
Comment (iamsh4shank) [inline app/src/main/java/org/mifos/mobile/ui/activities/PassCodeActivity.java:20]: Leave a line space after this.

File: app/src/main/java/org/mifos/mobile/ui/activities/SplashActivity.java [hunk 10]
```diff
@@ -5,13 +5,27 @@
...
+import org.mifos.mobile.ui.enums.ChargeType;
+import org.mifos.mobile.ui.fragments.ClientAccountsFragment;
+import org.mifos.mobile.ui.fragments.ClientChargeFragment;
+import org.mifos.mobile.ui.fragments.RecentTransactionsFragment;
+import org.mifos.mobile.ui.fragments.ThirdPartyTransferFragment;
 import org.mifos.mobile.utils.Constants;
 
+import java.util.ArrayList;
+import java.util.List;
+
```
Comment (iamsh4shank) [inline app/src/main/java/org/mifos/mobile/ui/activities/SplashActivity.java:17]: All these imports are required?

File: app/src/main/java/org/mifos/mobile/ui/activities/HomeActivity.java [hunk 11]
```diff
@@ -6,7 +6,11 @@
 import android.content.DialogInterface;
 import android.content.Intent;
 import android.content.IntentFilter;
+import android.content.pm.ShortcutInfo;
+import android.content.pm.ShortcutManager;
 import android.graphics.Bitmap;
+import android.graphics.drawable.Icon;
+import android.net.Uri;
```
Comment (devansh-299) [inline app/src/main/java/org/mifos/mobile/ui/activities/HomeActivity.java]: remove this as this import is unused

File: app/src/main/java/org/mifos/mobile/ui/activities/HomeActivity.java [hunk 12]
```diff
@@ -45,13 +49,23 @@
 
 import javax.inject.Inject;
 
+import androidx.annotation.NonNull;
```
Comment (devansh-299) [inline app/src/main/java/org/mifos/mobile/ui/activities/HomeActivity.java]: same here, unused import

File: app/src/main/java/org/mifos/mobile/ui/activities/HomeActivity.java [hunk 13]
```diff
@@ -45,13 +49,23 @@
...
+import androidx.annotation.Nullable;
```
Comment (devansh-299) [inline app/src/main/java/org/mifos/mobile/ui/activities/HomeActivity.java]: this also

File: app/src/main/java/org/mifos/mobile/ui/activities/HomeActivity.java [hunk 14]
```diff
@@ -45,13 +49,23 @@
...
 import androidx.appcompat.app.ActionBarDrawerToggle;
 import androidx.core.content.ContextCompat;
 import androidx.core.view.GravityCompat;
 import androidx.drawerlayout.widget.DrawerLayout;
 import androidx.fragment.app.Fragment;
 import androidx.fragment.app.FragmentManager;
 import androidx.localbroadcastmanager.content.LocalBroadcastManager;
+
+import java.util.ArrayList;
+import java.util.Arrays;
```
Comment (devansh-299) [inline app/src/main/java/org/mifos/mobile/ui/activities/HomeActivity.java]: same goes for this line, 64, 65 and 67 line

File: app/src/main/java/org/mifos/mobile/ui/activities/PassCodeActivity.java [hunk 15]
```diff
@@ -4,6 +4,7 @@
 import android.content.DialogInterface;
 import android.content.Intent;
 import android.os.Bundle;
+import android.util.Log;
```
Comment (devansh-299) [inline app/src/main/java/org/mifos/mobile/ui/activities/PassCodeActivity.java]: remove this

File: app/src/main/java/org/mifos/mobile/ui/activities/SplashActivity.java [hunk 16]
```diff
@@ -5,13 +5,27 @@
  */
 
 import android.content.Intent;
+import android.content.pm.ShortcutInfo;
```
Comment (devansh-299) [inline app/src/main/java/org/mifos/mobile/ui/activities/SplashActivity.java]: remove the following imports as these aren't being used :- 8, 9 , 10, 12, 18, 19, 20, 21, 22, 23, 26, 27 (these are line numbers)

**Review by devansh-299 — CHANGES_REQUESTED** [review]

**Comment by iamsh4shank** [comment]
Great you added all this feature, but I have a one suggestion there is no uninstall and remove option, so add it also.

**Comment by nirmalhk7** [comment]
> Great you added all this feature, but I have a one suggestion there is no uninstall and remove option, so add it also.

@robustTechie [Android's documentation on shortcuts](https://developer.android.com/guide/topics/ui/shortcuts) states that we can implement only four shortcuts at the maximum. So I dont think that is possible without changing the 4 existing shortcuts implemented as well.

**Pull request closed without merging**